### PR TITLE
feat(RELEASE-1116): timestamp variable is for build date

### PR DESCRIPTION
**Pull request by johnbieren**
This commit modifies the apply-mapping task to treat the timestamp user provided variable as the build timestamp instead of the current time. It is found by inspecting the image's labels. The existing timestamp functionality now exists under the release_timestamp variable.

Also, if a tag is translated with a null value, the task will now fail.

### CODE REVIEW

File: tasks/apply-mapping/apply-mapping.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ kind: Task
 metadata:
   name: apply-mapping
   labels:
-    app.kubernetes.io/version: "1.5.1"
+    app.kubernetes.io/version: "1.6.0"
```
Comment (mmalina) [inline tasks/apply-mapping/apply-mapping.yaml:7]: No-op: One could argue that it's a breaking change. But I think this is ok. My expectation is that users won't notice or mind (except those that actually expressed the need for build-time timestamps).

File: tasks/apply-mapping/apply-mapping.yaml [hunk 2]
```diff
@@ -86,6 +89,13 @@ spec:
                 variable=$(jq -c --argjson i "$i" '.[$i]' <<< "${SUPPORTED_VARIABLES}")
                 KEY=$(jq -r 'to_entries[] | .key' <<< "$variable")
                 VALUE=$(jq -r 'to_entries[] | .value' <<< "$variable")
+                # If a tag is going to be replaced by a null value, error out
```
Comment (mmalina) [inline tasks/apply-mapping/apply-mapping.yaml:92]: It's probably good to have this check. But I can't think of any way this could happen. Can you? But maybe we saw a -null tag generated at some point? Not sure.

File: tasks/apply-mapping/apply-mapping.yaml [hunk 3]
```diff
@@ -86,6 +89,13 @@ spec:
...
+                if [[ "$VALUE" == "null" ]] && \
+                  [[ $(jq --arg key "$KEY" '. | map(select(test($key))) | length' <<< "$tags") -gt 0 ]] ; then
+                    echo "Requested substitution for key: $KEY but calculated value for it was: $VALUE"
```
Comment (mmalina) [inline tasks/apply-mapping/apply-mapping.yaml]: Optional: This check is specifically for "null", so I think just printing "null" directly would imo be more straightforward. But if you prefer this, it's fine.

File: tasks/apply-mapping/apply-mapping.yaml [hunk 4]
```diff
@@ -186,11 +196,22 @@ spec:
             build_sha=$(echo "$IMAGE_REF" | cut -d ':' -f 2)
             passedTimestampFormat=$(jq -r --arg default "$defaultTimestampFormat" \
               '.timestampFormat // $default' <<< "$component")
-            timestamp="$(date -d "$currentTimestamp" "+$passedTimestampFormat")"
+            release_timestamp="$(date -d "$currentTimestamp" "+$passedTimestampFormat")"
+            arch_json="$(get-image-architectures "${IMAGE_REF}")"
+            # The build-date label is not the same per architecture, but we don't support separate tags per arch. So,
+            # we just use inspect the first digest listed
```
Comment (mmalina) [inline tasks/apply-mapping/apply-mapping.yaml]: Seems like "inspect" should be removed?

File: tasks/apply-mapping/apply-mapping.yaml [hunk 5]
```diff
@@ -186,11 +196,22 @@ spec:
...
+            arch="$(jq -rs 'map(.platform.architecture) | .[0]' <<< "$arch_json")"
+            os="$(jq -rs 'map(.platform.os) | .[0]' <<< "$arch_json")"
+            build_date="$(skopeo inspect --no-tags --override-os "${os}" --override-arch "${arch}" \
+              docker://"${IMAGE_REF}" | jq -r '.Labels."build-date"')"
+            timestamp="$(date -d "${build_date}" "+$passedTimestampFormat")"
```
Comment (mmalina) [inline tasks/apply-mapping/apply-mapping.yaml:207]: Nice! I thought it might be hard to parse the date string, but it seems `-d` is able to parse anything as long as it's a reasonable date string. Cool.

File: tasks/apply-mapping/apply-mapping.yaml (hunk 2 above)
Comment (mmalina) [reply tasks/apply-mapping/apply-mapping.yaml:92]: Now I see that you explicitly set timestamp to null if we can't get the build-date. So yeah, I guess it makes sense. But my question still stands.

**Review by mmalina — COMMENTED** [review]
It looks good. Added some comments, but no real issues.

File: tasks/apply-mapping/apply-mapping.yaml (hunk 2 above)
Comment (johnbieren) [reply tasks/apply-mapping/apply-mapping.yaml:92]: This can also happen to git_sha if `source.git.revision` is not there in the component. I *think* someone hit that a long time ago

File: tasks/apply-mapping/apply-mapping.yaml (hunk 1 above)
Comment (johnbieren) [reply tasks/apply-mapping/apply-mapping.yaml:7]: I left it as is since I didn't add or remove a task parameter, but I don't really have a strong opinion on it

File: tasks/apply-mapping/apply-mapping.yaml (hunk 3 above)
Comment (johnbieren) [reply tasks/apply-mapping/apply-mapping.yaml]: I updated it

File: tasks/apply-mapping/apply-mapping.yaml (hunk 4 above)
Comment (johnbieren) [reply tasks/apply-mapping/apply-mapping.yaml]: Nice catch, fixed

File: tasks/apply-mapping/apply-mapping.yaml (hunk 1 above)
Comment (mmalina) [reply tasks/apply-mapping/apply-mapping.yaml:7]: Yep, I think it's fine. I was just thinking out loud :)

File: tasks/apply-mapping/apply-mapping.yaml (hunk 2 above)
Comment (mmalina) [reply tasks/apply-mapping/apply-mapping.yaml:92]: ok

**Review by mmalina — APPROVED** [review]
